sync15-adapter/src/bso_record.rs

```diff
@@ -117,21 +131,97 @@ impl<T> DerefMut for BsoRecord<T> {
     }
 }
 
-impl<T> BsoRecord<T> {
-    /// If T is a Sync15Record, then you can/should just use record.into() instead!
+/// Represents the decrypted payload in a Bso. Provides a minimal layer of type safety to avoid double-encrypting.
+///
+/// Note: If we implement a full sync client in rust we may want to consider using stronger types for each record
+/// (we did this in the past as well), but for now, since everything is just going over the FFI, there's not a lot of
+/// benefit here.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Payload {
+    pub id: String,
+
+    #[serde(default)]
+    #[serde(skip_serializing_if = "is_false")]
+    pub deleted: bool,
+
+    #[serde(flatten)]
+    pub data: Map<String, JsonValue>,
+}
+
+// `#[serde(skip_if)]` only allows a function (not an expression).
+// Is there a builtin way to do this?
+#[inline]
+fn is_false(b: &bool) -> bool {
+    !*b
+}
+
+impl Payload {
+
     #[inline]
-    pub fn new_non_record<I: Into<String>, C: Into<String>>(id: I, coll: C, payload: T) -> BsoRecord<T> {
-        BsoRecord {
-            id: id.into(),
-            collection: coll.into(),
-            ttl: None,
-            sortindex: None,
-            modified: ServerTimestamp::default(),
-            payload,
+    pub fn new_tombstone(id: String) -> Payload {
+        Payload { id, deleted: true, data: Map::new() }
+    }
+
+    #[inline]
+    pub fn id(&self) -> &str {
+        &self.id[..]
+    }
+
+    #[inline]
+    pub fn is_tombstone(&self) -> bool {
+        self.deleted
+    }
+
+    pub fn into_bso(
+        self,
+        collection: String
+    ) -> CleartextBso {
+        let id = self.id.clone();
+        CleartextBso {
+            id,
+            collection,
+            modified: 0.0.into(), // Doesn't matter.
+            sortindex: None, // Should we let consumer's set this?
+            ttl: None, // Should we let consumer's set this?
+            payload: self,
         }
     }
+
+    pub fn from_json(value: JsonValue) -> error::Result<Payload> {
+        Ok(serde_json::from_value(value)?)
+    }
+
+    pub fn into_record<T>(self) -> error::Result<T> where for<'a> T: Deserialize<'a> {
+        Ok(serde_json::from_value(JsonValue::from(self))?)
+    }
+
+    pub fn from_record<T: Serialize>(v: T) -> error::Result<Payload> {
+        // TODO: This is dumb, we do to_value and then from_value. If we end up using this
+        // method a lot we should rethink... As it is it should just be for uploading
```

Comment:
Basically translating a struct into a map? (I think keeping it as you have it now makes sense; I'm guessing this'll change if we decide to use strongly-typed records, anyway?)